core-processor/src/main/java/io/micronaut/expressions/parser/ast/access/ContextMethodCall.java

```diff
@@ -0,0 +1,108 @@
+import org.objectweb.asm.commons.Method;
+
+import java.util.List;
+
+import static io.micronaut.expressions.parser.ast.types.TypeDescriptors.EVALUATED_EXPRESSION_TYPE;
+import static io.micronaut.expressions.parser.ast.util.EvaluatedExpressionCompilationUtils.getRequiredClassElement;
+import static org.objectweb.asm.Opcodes.CHECKCAST;
+
+/**
+ * Expression node used for invocation of method from expression
+ * evaluation context.
+ *
+ * @author Sergey Gavrilov
+ * @since 4.0.0
+ */
+@Internal
+public class ContextMethodCall extends AbstractMethodCall {
+
+    protected static final Method GET_BEAN_METHOD =
+        new Method("getBean", Type.getType(Object.class),
+            new Type[]{Type.getType(Class.class)});
+
+    public ContextMethodCall(String name, List<ExpressionNode> arguments) {
+        super(name, arguments);
+    }
+
+    @Override
+    protected CandidateMethod resolveUsedMethod(ExpressionCompilationContext ctx) {
+        List<Type> argumentTypes = resolveArgumentTypes(ctx);
+
+        ExpressionEvaluationContext evaluationContext = ctx.evaluationContext();
+        List<CandidateMethod> candidateMethods =
+            evaluationContext.getMethods(name)
+                .stream()
+                .map(method -> toCandidateMethod(ctx, method, argumentTypes))
+                .filter(method -> method.isMatching(ctx.visitorContext()))
+                .toList();
+
+        if (candidateMethods.isEmpty()) {
+            throw new ExpressionCompilationException(
+                "No method [ " + name + stringifyArguments(ctx) + " ] available in evaluation context");
+        } else if (candidateMethods.size() > 1) {
+            throw new ExpressionCompilationException(
+                "Ambiguous expression evaluation context reference. Found " + candidateMethods.size() +
+                    " matching methods: " + candidateMethods);
+        }
+
+        return candidateMethods.iterator().next();
+    }
+
+    @Override
+    public void generateBytecode(ExpressionCompilationContext ctx) {
+        GeneratorAdapter mv = ctx.methodVisitor();
+        Type calleeType = usedMethod.getOwningType();
+
+        ClassElement calleeClass = getRequiredClassElement(calleeType, ctx.visitorContext());
+
+        pushGetBeanFromContext(mv, calleeType);
+        compileArguments(ctx);
+        if (calleeClass.isInterface()) {
+            mv.invokeVirtual(calleeType, usedMethod.toAsmMethod());
+        } else {
+            mv.invokeVirtual(calleeType, usedMethod.toAsmMethod());
+        }
+    }
+
+    /**
+     * Pushing method obtaining bean of provided type from beanContext.
+     *
+     * @param mv       methodVisitor
+     * @param beanType required bean typ
+     */
+    protected void pushGetBeanFromContext(GeneratorAdapter mv, Type beanType) {
+        mv.loadThis();
+        mv.push(beanType);
+
+        // invoke getBean method
+        mv.invokeVirtual(EVALUATED_EXPRESSION_TYPE, GET_BEAN_METHOD);
```

Comment:
to get good error messages it may be necessary to also passing the `BeanResolutionContext` depending where this expression is evaluated.

Reply:
I've kept in mind that expressions in annotation metadata can be evaluated separately from instantiation of beans owning that annotation metadata. So instead of directly passing `BeanResolutionContext` to `getBean` method of `AbstractEvaluatedExpression` I've implemented the logic according to which each evaluated expression class maintains it's own `BeanResolutionContext` which it passes when obtains beans from `BeanContext`. At least this approach allows to solve circular dependency issues. You can take a look at `AbstractEvaluatedExpression#getBean` method